packages/playground/src/playground-db.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+defined('ABSPATH') || exit;
+
+function playground_escape_array($array)
+{
+	global $wpdb;
+	$escaped = array();
+	foreach ($array as $value) {
+		if (is_numeric($value))
+			$escaped[] = $wpdb->prepare('%d', $value);
+		else
+			$escaped[] = $wpdb->prepare('%s', $value);
+	}
+	return implode(',', $escaped);
+}
+
+function playground_dump_db($zip)
+{
+	global $wpdb;
+
+	$tables   = playground_get_db_tables();
+	$sql_dump = array();
+
+	foreach ($tables as $table) {
+		array_push(
+			$sql_dump,
+			sprintf("DROP TABLE IF EXISTS `%s`;", esc_sql($table)),
```

Comment:
What's the difference between `esc_sql` and `$wpdb->prepare()`? Would it make sense to stick to one of these?

Reply:
`$wpdb->prepare()` will add quotes around table name, and it's a practice in WP to not add table names with `$wpdb->prepare()`.

Reply:
Aha, gotcha! Is `sprintf()` typically used in WordPress core in cases like that? If yes, we're good. If not, is it possible  do here what core does?

Reply:
You would usuall see `"SELECT * FROM $wpdb->posts"` but this would require us to save the escaped value in a variable, or contact strings `'DROP TABLE IF EXISTS `' . esc_sql($table) . '`;'` I think that `sprintf` is a ok solution here.

Reply:
The [doc page for esc_sql](https://developer.wordpress.org/reference/functions/esc_sql/) says this:

> Be careful in using this function correctly. It will only escape values to be used in strings in the query. That is, it only provides escaping for values that will be within quotes in the SQL (as in field = '{$escaped_value}'). If your value is not going to be within quotes, your code will still be vulnerable to SQL injection. For example, this is vulnerable, because the escaped value is not surrounded by quotes in the SQL query: ORDER BY {$escaped_value}. As such, this function does not escape unquoted numeric values, field names, or SQL keywords.

Indeed it escapes backslashes but doesn't seem to properly escape backticks:

```php
<?php
echo "`" . esc_sql("A backtick ` and a backslash " . chr(92)) . "`";
// `A backtick ` and a backslash \\`
```

Here's what `$wpdb->quote_identifier` does:

```php
<?php
echo $wpdb->quote_identifier("A backtick ` and a backslash " . chr(92))
// `A backtick `` and a backslash \`
```

The latter seems to be correct:

```sql
create table `te``st\` (id int);
show tables;
```

```
Tables_in_test
--
te`st\
```

Reply:
In that case there is no point in escaping table names, which should be fine as they are coming from the DB.
Should I just do `"SELECT * FROM $table"`?

Reply:
Well, in the (highly unlikely) case I listed above, `"SELECT * FROM $table"` would not work, as it would become `"SELECT * FROM te`st\"` but `sprintf("SELECT * FROM %s", $wpdb->quote_identifier($table))` would work as it would become `"SELECT * FROM `te``st\`"` which is valid MySQL SQL.

Reply:
I see now. What about esc_sql?
It seems like it's not adding any security so we can remove it.

Reply:
Sorry about all the questions I would just like to ensure we are going in the same direction.

Reply:
Agreed about removing `esc_sql`. `sprintf("SELECT * FROM `%s`", esc_sql($table))` would return invalid SQL:

```
SELECT * FROM `te`st\\`
```

`$wpdb->quote_identifier` looks like a good replacement for `esc_sql`

Reply:
I updated this in c02934d36a2fe2dec4ad6d871179182c3a8f03e6.
Thank you for your help on this. It took a while, but I think that we landed on a solid implementation.